client/json-rpc/src/async_client/client.rs

```diff
@@ -280,53 +376,121 @@ impl<R: RetryStrategy> Client<R> {
         }
     }
 
-    pub async fn send_without_retry(
+    pub async fn send_without_retry(&self, request: &Request) -> Result<JsonRpcResponse, Error> {
+        let rpc_resp: JsonRpcResponse = self.send_json_request(request.to_json(1)).await?;
+        self.validate(rpc_resp)
+    }
+
+    /// Batch requests into one JSON-RPC batch request.
+    /// Batch responses may contain partial success response, or different errors,
+    /// hence this function won't be able to handle errors for caller, including
+    /// retry on specific error.
+    ///
+    /// Caller should handle errors and retry as needed.
+    ///
+    /// After errors are handled, caller can convert `JsonRpcResponse` into `Response<T>`
+    /// in order of requests:
+    ///
+    /// ```rust
+    /// use libra_json_rpc_client::async_client::{
+    ///     types as jsonrpc, Client, Error, Request, Response, Retry,
+    /// };
+    /// use std::convert::{TryInto};
+    ///
+    /// # async fn doc() -> Result<(), Error> {
+    /// let client = Client::from_url("http://testnet.libra.org/v1", Retry::default()).unwrap();
+    /// let mut res = client
+    ///     .batch_send(vec![Request::get_metadata(), Request::get_currencies()])
+    ///     .await?;
+    ///
+    /// let metadata: Response<jsonrpc::Metadata> = res
+    ///     .remove(0)?
+    ///     .try_into()?;
+    ///
+    /// let currencies: Response<jsonrpc::CurrencyInfo> = res
+    ///     .remove(0)?
+    ///     .try_into()?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    ///
+    pub async fn batch_send(
+        &self,
+        requests: Vec<Request>,
+    ) -> Result<Vec<Result<JsonRpcResponse, Error>>, Error> {
+        self.batch_send_without_retry(&requests).await
+    }
+
+    pub async fn batch_send_without_retry(
         &self,
-        method: &str,
-        params: &serde_json::Value,
-    ) -> Result<JsonRpcResponse, Error> {
+        requests: &[Request],
+    ) -> Result<Vec<Result<JsonRpcResponse, Error>>, Error> {
+        let json_requests: Vec<serde_json::Value> = requests
+            .iter()
+            .enumerate()
+            .map(|(i, r)| r.to_json(i))
+            .collect();
+
+        let rpc_resps: Vec<JsonRpcResponse> = self.send_json_request(json![json_requests]).await?;
+
+        let mut result: Vec<Result<JsonRpcResponse, Error>> = json_requests
+            .iter()
+            .map(|r| {
+                Err(Error::unexpected_response_not_found_error(format!(
+                    "request: {}",
+                    r
+                )))
+            })
+            .collect::<Vec<_>>();
+        for resp in rpc_resps {
+            if let Some(ref id) = resp.id {
+                if let Ok(index) = serde_json::from_value::<usize>(id.clone()) {
+                    if index < result.len() {
+                        result[index] = self.validate(resp);
```

Comment:
it is pretty dangerous to use response id as index to array, also that's why you needed this check,  I would just change the outer loop instead

for (i, resp) in rpc_resps.enumeration() { 

  // check i == index
result[i] = self.validate(resp) 
}

Reply:
yeah, in protocol the response could be randomly ordered, I think we probably should use an hash map (id as key) instead.

Reply:
it's same.
the primary thing here is what is request id, as interface accept an array of requests, we should expect response is an array and index matches.
anything does not match request id, should not be put in result, although, maybe put into result is better for checking later, but I don't think this is something we really need care too much, as if server side has bug causes response id != request id, it will be a huge problem and basically you can't use batch send anymore.

Reply:
also note the result is pre-init with Err(response not found error with request data), hence no response in result for a specific request will report error

Reply:
I think the question here is when one response has invalid id, do we fail the whole batch send call, or we return valid ones and ignore responses that have invalid/mismatched ids.
Either way is fine to me.

Reply:
IMHO to be on the safe side, we should really start with an all-or-nothing approach ..

be careful if the response contains a invalid id (could be anything) , you don't want it to trigger client crash though.